Add dag-pb investigation doc

rvagg opened the pull request:
Readable version @ https://github.com/ipld/ipld/blob/rvagg/dag-pb-investigation/design/tricky-choices/dag-pb-forms-impl-and-use.md

This relates to https://github.com/ipfs/go-ipfs/pull/7976 and is intended to help with some decision making, or at a minimum, to help build a brief summary of changes for users.

There's a lot in here but for those who are deep in this level of the stack I think it'll be informative, and we have some choices to make coming out of this.

@warpfork @aschmahmann @Stebalien @achingbrain I think you might want to all cast an eye over this. Sorry that it's so long, but there's a lot of history and implementation detail being covered! There's a couple of options outlined at the bottom for moving forward, but I'm sure there are other options we could come up with.

willscott commented:
I'm going to be a proponent of 'option 1'

the filecoin risk and the IPFS work proposed in option 2 seem not worth it to me

olizilla commented on design/tricky-choices/dag-pb-forms-impl-and-use.md:
```diff
@@ -0,0 +1,126 @@
+
+## Go data forms
+
+go-ipfs defers all its dag-pb concerns to [go-merkledag](https://github.com/ipfs/go-merkledag/). Which in turn serves as a translation layer between the serialized forms and the in-memory forms of these nodes. Serialization happens in [go-merkledag/pb](https://github.com/ipfs/go-merkledag/tree/master/pb) which started life as a protobuf codegen but has been manually tweaked since then my a few of ...
```
```suggestion
go-ipfs defers all its dag-pb concerns to [go-merkledag](https://github.com/ipfs/go-merkledag/). Which in turn serves as a translation layer between the serialized forms and the in-memory forms of these nodes. Serialization happens in [go-merkledag/pb](https://github.com/ipfs/go-merkledag/tree/master/pb) which started life as a protobuf codegen but has been manually tweaked since then by a few of us for various reasons.
```

olizilla commented on design/tricky-choices/dag-pb-forms-impl-and-use.md:
```diff
@@ -0,0 +1,126 @@
+* js-ipfs needs to stop forcing dag-pb results from `dag get` back through the dag-pb codec, just output what is resolved in the codec requested
+* js-ipfs needs other new goodness introduced in go-ipld-prime in go-ipfs
+
+### Option 2: We go the data model form wrong
```
```suggestion
### Option 2: We got the data model form wrong
```

olizilla commented on design/tricky-choices/dag-pb-forms-impl-and-use.md:
```diff
@@ -0,0 +1,126 @@
+
+In https://github.com/ipld/specs/issues/245 and then https://github.com/ipld/specs/pull/247 an IPLD Schema form of dag-pb was merged. There was some discussion about casing of names at https://github.com/ipld/specs/pull/247#discussion_r390220519 where it was resolved that we should match the protobuf schema. We were apparently oblivious to the additional `ProtoNode` representation, _or_ we didn't ...
+
+The IPLD Schema has since been used to rewrite the new JS and Go codecs for dag-pb, but this was not a major departure since these codecs (originally js-ipld-dag-pb and go-merkledag/pb) at their protobuf level were always using these forms. We were adding additiona lenses onto these codecs before they presented themselves at the IPFS `dag` API - via `ProtoNode`, `ipld.Link` and the custom printin ...
```
```suggestion
The IPLD Schema has since been used to rewrite the new JS and Go codecs for dag-pb, but this was not a major departure since these codecs (originally js-ipld-dag-pb and go-merkledag/pb) at their protobuf level were always using these forms. We were adding additional lenses onto these codecs before they presented themselves at the IPFS `dag` API - via `ProtoNode`, `ipld.Link` and the custom printing forms in js-ipfs `dag get`.
```

aschmahmann commented on design/tricky-choices/dag-pb-forms-impl-and-use.md:
```diff
@@ -0,0 +1,126 @@
+5. The custom `LookupByString` means that you can't resolve anything else in the data model form of the dag-pb node.
+6. There is an `/ipld/` override for resolution, such that when the path is prefixed with this, will _not_ perform the unixfsnode reification and you'll get the "pure" data model form and can resolve all of the properties of `{Data,Links:[{Name,Tsize,Hash},...]}` in a standard pathing form.
+
+Worth noting is that the current (pre go-ipld-prime) version of go-ipfs **does not do anything special with the `/ipld/` path prefix and dag-pb nodes**. It appears to behave the same as without, resolving named links and nothing else.
```
This is close, but I don't think quite accurate. IIUC what's happening both pre and post go-ipld-prime with respect to /ipfs and /ipld is the following:

Pre go-ipld-prime:
- /ipfs -> Interpret the data as UnixFS, meaning QmFoo/NamedItemInAHAMTDirectory/bar` works
- /ipld -> Interpret the data as non-IPLD, but with ["legacy pathing"](https://github.com/ipld/specs/blob/a652a23ba974c7b524882602a12b6d978b13829a/block-layer/codecs/dag-pb.md#alternative-legacy-pathing) that traverses the named links and hides the size and data fields. In go-ipld-prime world this is analogous to using [PathPBNode](https://github.com/ipfs/go-unixfsnode/blob/d6733f6db69431075c9cf7b30e0ee1f6a082f4b0/pathpbnode.go) as a lens.

Post go-ipld-prime (branch today):
- /ipfs -> Interpret the data as UnixFS
- /ipld -> Interpret the data as plain as possible without any legacy pathing

Main takeaways from an investigative view (i.e. not judging the outcomes):
- /ipld did used to do something in go-ipfs, go-ipld-prime is changing the behavior
- Keeping /ipfs is not a reliable way for us to do named pathing through DagPB objects since we'll use the UnixFS children instead of the object links (e.g. no accessing of named HAMT shard links). Additionally, if we add more strictness to the UnixFS reifier then even basic non-UnixFS DagPB nodes might be untraversable this way.
- As it stands with go-ipld-prime we have killed off named path traversal of DagPB objects via `ipfs dag get`

rvagg replied:
I don't quite understand the last two points you have here @aschmahmann.

From what I understand, and from experimenting, it looks like `QmFoo/NamedLink/foo` _and_ `QmFoo/NamedShardedLink/foo` works old and new, it's just that the output is now dag-json and has the TitleCased `Data` and `Links`. Aside from that, the difference is in `/ipld`, as you've pointed out, works for named links with the old, but not sharded named links. I can do `/ipld/QmFoo/NamedLink` but not `/ipld/QmFoo/NamedShardedLink` but I can do `/ipld/QmFoo/AA`, `AA` being one of the names that the HAMT is going to give for its sharded named links.

So with the go-ipld-prime changes, the only loss (aside from json changes) is the ability to use unsharded named links on sharded directories (`AA` etc.). But you still have the ability to access those by index, so `/ipld/QmFoo/Links/170` is going to get me the same as `/ipld/QmFoo/AA` would have got me before.

Which is all pretty obscure and doesn't seem like a loss of functionality for the vast majority of users, likely no users since this is all hidden (including behind a flag in go-ipfs!).

I'm going to update the doc to talk about sharded directories because that's a difference I didn't investigate yesterday.

olizilla reviewed (commented):
Good digging! i mearly offer some typo fixes.

> the `data` property is going to end up as an ugly associative array: `{"data":{"0":115,"1":111,"2":109,"3":101,"4":32,"5":100,"6":97,"7":116,"8":97},"links":[...]}`

is a familiar sight from hacking on explore.ipld.io

BigLep commented:
Thanks for the write up.  I read through it, but I know I'm not catching all the implications.  My hope is that we end in a world that is simpler/more intuitive.  It sounds like there is followup work here depending on the option.  Please make sure the corresponding issues get created/linked.

rvagg commented:
Added some additional notes about sharding and `/ipld/` semantics in the latest commit. Also added a slightly rambly section at the end with "Option 3" that's a kind of compromise but IMO is worst of both worlds. I wanted to introduce it here because ADLs, which we are already using via go-unixfsnode, provide us with a way to morph the data model into new forms which we could do to go from `Data`, `Links` and friends to `data`, `links` and friends. But .. I think we'd just end up creating more problems for ourselves and confusing users even more as they look behind each successive curtain.

aschmahmann commented:
> "Option 3" that's a kind of compromise but IMO is worst of both worlds

Agreed. Option 1 seems like the most sane thing to do. It continues to give us UnixFS support via `/ipfs` and non-magical and understandable pathing support via `/ipld`. The edge cases around named DagPB links are a bit of a shame, but they were pretty confusing anyway and AFAICT not particularly used.

warpfork commented:
A lot of digital ink has already been spilled on this, so I'll keep my comment on choices here to a short "2 cents":

- I think the TitleCase is unfortunate, and weird.  It's a golang naming convention that escaped to affect our datamodel/json, and if we were in a greenfield, I'd say that's bad.
- That said, I frankly don't have a lot of spoons for caring about dag-pb.  And we're not in a greenfield.  So it can be weird all it wants.  And I'm on board with work-minimization.
- The rest of the remarks in Option 1 about moving towards consistency in datamodel pathing vs named links pathing checks out to me.  :white_check_mark:
